### ISSUE TITLE
Use generics to reduce duplicated code in the IO HDF5 module

### ISSUE THREAD

**Issue by abensonca** [galacticusorg/galacticus#94]
Read/write functions for different types and ranks should be collapsible into single generic functions.

**Comment by abensonca** [comment]
Depends on #92

**Comment by abensonca** [comment]
For many of the functions we don't need generics - using assumed rank arrays for the arguments would let a single function handle dataset writes (for example) for all possible ranks.

**Comment by abensonca** [comment]
# Analysis: recommended approach

**Recommendation: a hybrid — the `<generic>` preprocessor template over the *type* axis, and Fortran 2018 assumed-rank (`dimension(..)`) over the *rank* axis.** This is exactly the pattern already proven in `source/utility/polyranks.F90`, and neither mechanism alone captures the full win.

## Duplication structure

`source/utility/IO/HDF5/_module.F90` is 18,340 lines (the largest file in the repo) with ~90 near-duplicate procedures spanning {integer, integer8, double, character, varying_string, var-length} × {rank 1–6} × {allocatable/static} × {read/write} × {dataset/attribute}. Comparing variants:

- **Across ranks**, only 3–15 lines differ per ~170-line procedure: the `dimension(:,:,...)` declarations, the `datasetRank` literal, the `datasetDimensions` extent, and the append-dimension filtering (trivial for 1D). The `h5dwrite_f`/`h5dread_f` calls themselves are **identical across all ranks**.
- **Across types**: the `H5T_*` constant, the `c_loc` + contiguous-copy machinery for `kind_int8`, and `h5tcopy`/`h5tset_size` datatype setup for character.
- **Allocatable vs. static reads** differ by ~5 lines out of ~370 (allocate vs. bounds check). There is already copy-paste drift as evidence of the maintenance cost — e.g. a duplicated open-check block at lines 9598 and 9603.

## Why each axis gets a different tool

**Rank → assumed-rank**, because it eliminates variants at both the source *and* object-code level: one compiled procedure per type instead of six (also helping compile time on this file). gfortran ≥ 16 gives full F2018 support, and `select rank` is already in use (`polyranks.F90`, `numerical/interpolation/1D.F90`, `multi_dimensional_minimizer.F90`). With HDF5 ≥ 1.14.5 as the floor, the library's Fortran generics are themselves `type(*), dimension(..)`-based (verified against the 1.14 modules), so assumed-rank buffers pass **directly** to `h5dwrite_f`/`h5dread_f` — no `select rank` needed around the calls at all. The only `select rank` blocks required are the small allocation blocks in allocatable-read variants (the associating entity retains the `allocatable` attribute inside a `rank(n)` block, per the `polyranks.F90` constructor pattern).

**Type → the `<generic>` preprocessor**, because each type needs a different `H5T_*` constant and one cannot `select type` on a `type(*)` buffer. One directive with the H5T constant as an instance attribute:

```fortran
!![
<generic identifier="Type">
 <instance label="Integer"  intrinsic="integer"            h5type="H5T_NATIVE_INTEGER"/>
 <instance label="Integer8" intrinsic="integer(kind_int8)" h5type="H5T_INTEGER8"      />
 <instance label="Double"   intrinsic="double precision"   h5type="H5T_NATIVE_DOUBLE" />
</generic>
!!]

subroutine IO_HDF5_Write_Dataset_{Type¦label}(self,datasetValue,datasetName,...)
  {Type¦intrinsic}, dimension(..), intent(in   )                        :: datasetValue
  integer         (kind=HSIZE_T ), dimension(rank(datasetValue))       :: datasetDimensions
  ...
  call h5dwrite_f(...,{Type¦h5type},datasetValue,datasetDimensions,errorCode,...)
```

The `generic :: writeDataset => IO_HDF5_Write_Dataset_{Type¦label}` binding expands into the full list automatically (as in `dictionaries.F90`), so the public API and all call sites are untouched.

**Allocatable vs. static** needs neither mechanism: factor the shared ~360-line body into a private helper; the allocatable specific becomes a `select rank` allocate plus a call to it.

## Caveats

1. The `<allocate variable=... shape=.../>` build directive cannot handle assumed-rank variables (rank unknown at preprocess time) — use explicit `select rank` allocation blocks.
2. An assumed-rank specific is TKR-ambiguous with rank-specific ones of the same type, so each family is an atomic swap, not a gradual overlay. An assumed-rank dummy also accepts rank-0 actuals, so add a runtime rank check where scalars are invalid.
3. Character (custom datatype sizing), varying_string (already a thin wrapper), and the var-length variants (`VarDouble`, `VarVarDouble`, `VarInteger8` — element-wise `hdf5VlenC` marshaling) don't fit the template and should stay hand-written (~1.5–2k lines).

Back-of-envelope: the numeric families collapse from ~60 procedures to ~10 templated ones, taking the file from ~18.3k to roughly 6–8k lines.

## Folding in #92

With HDF5 ≥ 1.14.5 committed as the minimum, #92 should be folded into this work rather than sequenced before it — the template rewrite is the natural vehicle for most of #92's Tier B, and doing them separately means double churn (and double review) in the same central file:

- **int8 (datasets and attributes)**: better than #92's planned C→`c_ptr` conversion — with 1.14's assumed-type/assumed-rank interfaces, `integer(kind_int8)` buffers pass *directly* to `h5dwrite_f`/`h5aread_f` with the `H5T_INTEGER8` mem type. The C-API calls, `c_loc` wrapping, and contiguous copies are **deleted rather than converted**, and int8 joins the same template as integer/double (just a different `h5type` attribute).
- **VLEN**: these stay as hand-written specials, but get the `H5Dwrite`→`h5dwrite_f` `c_ptr` swap while being touched.
- **Region references** (27 `H5Dread` sites — the riskiest per #92): these are embedded in the read-dataset bodies being rewritten anyway, so the calling convention must be decided as part of the read-family rewrite regardless. Round-trip tests first.
- **Tier A** (`H5T_C_S1_Get`) rides along trivially.
- **Verified against the 1.14.5 Fortran modules**: there is still **no** `H5T_VARIABLE_F` constant (only `h5tis_variable_str_f`), so the 3-line `H5T_Variable_Get` helper should be kept for now; and `h5tbread_field_name_f` still has **no `integer(8)` specific** (only default integer, strings, and real kinds), so the Tier D `H5TBread_fields_name` wrapper also survives unless reimplemented via the core API — both out of scope, both cheap to keep.
- **Knock-on**: once the C-API interface block at `:388-460` is gone, the `herr_t` half of the `hdf5FCInterop` probe/rewrite machinery goes dead and can be retired in a final cleanup.

### Suggested sequencing

1. **PR 1**: round-trip tests in `tests.IO.HDF5.F90` (region refs, VLEN, int8 table column) + Tier A trivia — the safety net.
2. **PRs 2–n**: family-by-family template rewrite (attributes → numeric `writeDataset` → `readDataset`/`readDatasetStatic`), converting calling conventions in the same pass. Each step verifiable via the hdf5Diff workflow / byte-comparison against master.
3. **Final PR**: delete the C interface block, `hdf5_cTypes.c`/`hdf5_cFuncs.c` remnants (minus the two survivors above), and the dead interop machinery.

This inverts the previously noted "depends on #92" ordering: rather than sequencing, merge the two campaigns.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

**Comment by abensonca** [comment]
# Campaign complete

The deduplication proposed here is done, delivered as a stack of five PRs (merge in order; the last carries the closing keyword for this issue):

| PR | Scope | Net change |
|---|---|---|
| #1210 | Test safety net — round-trip coverage for **every** read/write variant (206→210 assertions) | +230 lines of tests |
| #1211 | Attribute family: 16 procedures → 4 templates | −1,077 |
| #1212 | Numeric `writeDataset`: 13 procedures → 1 assumed-rank template | −2,088 |
| #1213 | Numeric `readDataset`/`readDatasetStatic`: 21 procedures → 2 templates | −6,041 |
| #1214 | Final cleanup: VLEN 1-D pair + real/integer `readTable` → 3 templates | −631 |

**Result: `source/utility/IO/HDF5/_module.F90` shrinks from 18,340 to 8,473 lines (−54%), with 72 hand-written procedures collapsed into 11 templated implementations and zero call-site changes** — the `writeAttribute`/`readAttribute`/`writeDataset`/`readDataset`/`readTable` generic bindings are preserved exactly.

## Design as implemented

The hybrid approach from the analysis above, with one important correction:

- **Type axis** → `<generic>` preprocessor directives (`Type` for {integer, integer8, double}, plus `VlenType` and `TableType` for the cleanup pass), carrying the `H5T_*` constants, assert sets, and data-type enums as instance attributes.
- **Rank axis** → Fortran 2018 assumed-rank (`dimension(..)`) arguments, with `select rank` used only where the language requires it (allocating the allocatable read buffers).
- **Correction to the original plan:** HDF5 1.14.5's Fortran API does **not** provide assumed-type/assumed-rank buffer interfaces — its typed generics are kind×rank-specific (the generated `H5_gen` module). The unified calling convention is instead `c_loc()` on a `target, contiguous` assumed-rank dummy passed to the `c_ptr` interfaces (valid under TS29113; verified empirically with gfortran 16). A side benefit: all the int8 "allocate a contiguous copy" workarounds are gone, since `contiguous` makes the compiler handle copy-in/out.

## Capability gains (uniformity)

- Ranks 1–7 supported for all templated types in dataset reads/writes, and 0–7 for attributes (previously: integer stopped at 3-D, int8 at 4-D, double at 6-D; only double had a 2-D attribute).
- `appendDimension` available for all types and ranks (previously 2-D+ double and 2-D int8 only), with bounds validation.
- `readBegin`/`readCount`/`readSelection` available for all types and ranks, with new run-time validation that subsection arrays match the dataset rank.

## Latent bugs fixed along the way

- Duplicated open-object check in the double 1-D allocatable read.
- The 1-D allocatable `readSelection` path built its memory dataspace from `shape()` of the **not-yet-allocated** output array.
- Doubled `locationReport()` in integer scalar attribute read error messages.
- `Integer8_4D` write's inconsistent `commentText` argument name (normalized to `comment`; no keyword callers existed).

## The safety net mattered

PR #1210's tests caught a genuine would-be data-corruption regression during #1213: referenced reads into static arrays require the memory dataspace to be shaped like the caller's full buffer with a hyperslab selection (corner placement), not the contiguous layout the allocatable variant uses — a difference invisible at rank 1 and silently corrupting at rank ≥ 2.

## Relationship to #92

Most of #92 turned out to be already complete in-tree before this campaign began. The two verified survivors (checked against the HDF5 1.14.5 Fortran modules directly): `H5T_Variable_Get` (no `H5T_VARIABLE` constant exists in the Fortran API) and `H5TBread_fields_name` (the HL Fortran table API still has no `integer(8)` interface). Details in the #92 summary comment; #92 is also closed by #1214.

## Tooling

`Allocate.py` gained deferral for `<allocate>` directives whose target is declared with a leading type placeholder (`{Type¦intrinsic} :: x`), needed for allocation inside generic templates (#1211).

🤖 Generated with [Claude Code](https://claude.com/claude-code)
